### PR TITLE
[blobstore] Filter out expired blobs from query

### PR DESCRIPTION
**Pull request by ian-shim**
## Why are these changes needed?
Updates `GetBlobMetadataByStatusXXX` methods so that it doesn't fetch any expired blobs (determined by filtering on `RequestedAt > Now - TTL`). 

## Checks

- [ ] I've made sure the lint is passing in this PR.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by pschork — APPROVED** [review]

**Comment by ian-shim** [comment]
This works as long as we don't change the TTL, but this would break if TTL changes - which is defined onchain. 
I think the better solution is to either create an index based on `Expiry` attribute or delete the expired blobs. Any thoughts @dmanc ?

**Comment by dmanc** [comment]
> This works as long as we don't change the TTL, but this would break if TTL changes - which is defined onchain. I think the better solution is to either create an index based on `Expiry` attribute or delete the expired blobs. Any thoughts @dmanc ?

Why would we change the TTL? Also I don't like anything to do with assuming the metadata store doesn't have expired blobs. Blobs usually aren't deleted immediately when they get expired on DynamoDB for example, it goes through them at some rate.

["DynamoDB automatically deletes expired items within a few days of their expiration time, without consuming write throughput."](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html)

**Comment by ian-shim** [comment]
> Why would we change the TTL?

Unclear for now, but we may want to extend the TTL in the future. It's an assumption we have to keep in mind, so I'd rather handle it the right way.

> Also I don't like anything to do with assuming the metadata store doesn't have expired blobs.

I think thats fair. Should we create an index on `Expiry` then?

**Comment by dmanc** [comment]
> > Why would we change the TTL?
> 
> Unclear for now, but we may want to extend the TTL in the future. It's an assumption we have to keep in mind, so I'd rather handle it the right way.
> 
> > Also I don't like anything to do with assuming the metadata store doesn't have expired blobs.
> 
> I think thats fair. Should we create an index on `Expiry` then?

Yeah let's create an index

> > Why would we change the TTL?
> 
> Unclear for now, but we may want to extend the TTL in the future. It's an assumption we have to keep in mind, so I'd rather handle it the right way.
> 
> > Also I don't like anything to do with assuming the metadata store doesn't have expired blobs.
> 
> I think thats fair. Should we create an index on `Expiry` then?

Yeah I can add it tomorrow.

**Comment by dmanc** [comment]
Deployed the index, should be unblocked now

File: disperser/common/blobstore/blob_metadata_store.go [hunk 1]
```diff
@@ -117,10 +124,14 @@ func (s *BlobMetadataStore) GetBlobMetadataByStatus(ctx context.Context, status
 // Because this function scans the entire index, it should only be used for status with a limited number of items.
 // It should only be used to filter "Processing" status. To support other status, a streaming version should be implemented.
 func (s *BlobMetadataStore) GetBlobMetadataByStatusCount(ctx context.Context, status disperser.BlobStatus) (int32, error) {
```
Comment (jianoaix) [inline disperser/common/blobstore/blob_metadata_store.go:126]: This function name is a bit odd. It should probably be `GetBlobCountByStatus` or `GetBlobMetadatCountByStatus`.

File: disperser/common/blobstore/blob_metadata_store.go [hunk 2]
```diff
@@ -94,9 +98,12 @@ func (s *BlobMetadataStore) GetBlobMetadata(ctx context.Context, metadataKey dis
 // Because this function scans the entire index, it should only be used for status with a limited number of items.
 // It should only be used to filter "Processing" status. To support other status, a streaming version should be implemented.
 func (s *BlobMetadataStore) GetBlobMetadataByStatus(ctx context.Context, status disperser.BlobStatus) ([]*disperser.BlobMetadata, error) {
-	items, err := s.dynamoDBClient.QueryIndex(ctx, s.tableName, statusIndexName, "BlobStatus = :status", commondynamodb.ExpresseionValues{
+	items, err := s.dynamoDBClient.QueryIndex(ctx, s.tableName, statusIndexName, "BlobStatus = :status AND RequestedAt > :requestedAt", commondynamodb.ExpresseionValues{
 		":status": &types.AttributeValueMemberN{
 			Value: strconv.Itoa(int(status)),
+		},
+		":requestedAt": &types.AttributeValueMemberN{
```
Comment (jianoaix) [inline disperser/common/blobstore/blob_metadata_store.go:105]: `liveBlobBoundary` or something? More readable than two `requestedAt`s

**Review by jianoaix — COMMENTED** [review]
There seems an issue where Dispersing blobs that are expired will remain stuck in Dispersing after this change

File: disperser/common/blobstore/blob_metadata_store.go (hunk 1 above)
Comment (ian-shim) [reply disperser/common/blobstore/blob_metadata_store.go]: Yeah updated to `GetBlobMetadatCountByStatus`

**Comment by ian-shim** [comment]
> There seems an issue where Dispersing blobs that are expired will remain stuck in Dispersing after this change

That's true. Is it a concern?

**Comment by jianoaix** [comment]
> > There seems an issue where Dispersing blobs that are expired will remain stuck in Dispersing after this change
> 
> That's true. Is it a concern?

Not sure if there is unexpected access (and use) of these "zombie" blobs before they get removed, if so that can cause issue

**Comment by ian-shim** [comment]
> > > There seems an issue where Dispersing blobs that are expired will remain stuck in Dispersing after this change
> > 
> > 
> > That's true. Is it a concern?
> 
> Not sure if there is unexpected access (and use) of these "zombie" blobs before they get removed, if so that can cause issue

Added a check on `Expiry` and moves to `Failed` [if it has passed](https://github.com/Layr-Labs/eigenda/pull/692/files#diff-b3512a4819965cc0824e31e98c1add01712302a4740156af5b0b487106b59ab9R160). Wdyt?

**Comment by jianoaix** [comment]
> > > > There seems an issue where Dispersing blobs that are expired will remain stuck in Dispersing after this change
> > > 
> > > 
> > > That's true. Is it a concern?
> > 
> > 
> > Not sure if there is unexpected access (and use) of these "zombie" blobs before they get removed, if so that can cause issue
> 
> Added a check on `Expiry` and moves to `Failed` [if it has passed](https://github.com/Layr-Labs/eigenda/pull/692/files#diff-b3512a4819965cc0824e31e98c1add01712302a4740156af5b0b487106b59ab9R160). Wdyt?

Sg, thanks

**Review by jianoaix — APPROVED** [review]

**Pull request merged**